### PR TITLE
[WIP] Add HTLCs to channels

### PR DESCRIPTION
**Pull request by metalicjames**
This work in progress adds HTLCs to channels. Users can manually add an arbitrary number of HTLCs to their channel and later clear them by presenting the preimage or timing out the HTLC.

Adds two commands to `lit-af`: `add` and `clear`.

To add an HTLC to channel `1` of size `200000` sats with the given payment hash:
`add 1 200000 84ca18fd5458e3886bc0cabb58591240627dc333f925b7ea2d4cb5107e09db65`

To clear the HTLC with the payment preimage (where the HTLC became index 0):
`clear 1 0 fd9f9c832685de8342f845c56f946116`

(`SHA256(fd9f9c832685de8342f845c56f946116) = 84ca18fd5458e3886bc0cabb58591240627dc333f925b7ea2d4cb5107e09db65`)

To clear the HTLC via timeout:
`clear 1 0 00000000000000000000000000000000`

Still to do:
- [x] Add HTLC to channel state
- [x] Clear HTLC from channel state via preimage
- [x] Clear HTLC from channel state via timeout
- [x] Claim on-chain HTLC outputs after channel break
- [x] Handle channel state update collisions
    - [x] PreimageSig     -      PreimageSig
    - [x] HashSig            -      HashSig
    - [x] PreimageSig     -      HashSig
    - [x] DeltaSig            -      PreimageSig
    - [x] DeltaSig            -      HashSig
    - [x] DeltaSig           -      DeltaSig 
- [x] Custom HTLC timeouts 
- [ ] Check we're checking for all the needed sanity/security predicates
- [ ] Cleanup/refactor duplicated code 
- [x] Automatically clear all HTLCs we learn the preimage for 
- [x] Fix unit tests affected by changes

### CODE REVIEW

**Comment by delbonis** [comment]
That's a *lot* of commits.

**Comment by gertjaap** [comment]
Awesome work @metalicjames . I pulled your changes, compiled and tested. All functionality works as expected for me

File: cmd/lit-af/chancmds.go [hunk 1]
```diff
@@ -96,6 +96,23 @@ var historyCommand = &Command{
 	ShortDescription: "Show all the metadata for justice txs.\n",
 }
 
+var addHTLCCommand = &Command{
+	Format: fmt.Sprintf("%s%s%s\n", lnutil.White("add"), lnutil.ReqColor("channel idx", "amount", "RHash"), lnutil.OptColor("data")),
+	Description: fmt.Sprintf("%s\n%s\n",
+		"Add an HTLC of the given amount (in satoshis) to the given channel.",
+		"Optionally, the push operation can be associated with a 32 byte value hex encoded."),
```
Comment (gertjaap) [inline cmd/lit-af/chancmds.go]: `the push operation` = `the add HTLC operation` ?

File: cmd/lit-af/chancmds.go [hunk 2]
```diff
@@ -96,6 +96,23 @@ var historyCommand = &Command{
...
+	ShortDescription: "Add HTLC of the given amount (in satoshis) to the given channel.\n",
+}
+
+var clearHTLCCommand = &Command{
+	Format: fmt.Sprintf("%s%s%s\n", lnutil.White("clear"), lnutil.ReqColor("channel idx", "HTLC idx", "R"), lnutil.OptColor("data")),
+	Description: fmt.Sprintf("%s\n%s\n%s\n",
+		"Clear an HTLC of the given index from the given channel.",
+		"Optionally, the push operation can be associated with a 32 byte value hex encoded.",
```
Comment (gertjaap) [inline cmd/lit-af/chancmds.go]: `the push operation` = `the clear HTLC operation` ?

File: cmd/lit-af/shell.go [hunk 3]
```diff
@@ -108,6 +108,16 @@ func (lc *litAfClient) Shellparse(cmdslice []string) error {
 		return parseErr(err, "push")
 	}
 
+	if cmd == "add" {
```
Comment (gertjaap) [inline cmd/lit-af/shell.go]: wonder if `add` / `clear` is not too generic, and should be `addhtlc` and `clearhtlc` in stead.

File: litrpc/chancmds.go [hunk 4]
```diff
@@ -455,3 +455,133 @@ func (r *LitRPC) DumpPrivs(args NoArgs, reply *DumpReply) error {
...
+	HTLCIndex  uint32
+}
+
+func (r *LitRPC) AddHTLC(args AddHTLCArgs, reply *AddHTLCReply) error {
+	if args.Amt > consts.MaxChanCapacity || args.Amt < consts.MinOutput {
+		return fmt.Errorf(
+			"can't add HTLC %d max is 1 coin (100000000), min is 1", args.Amt)
+	}
+
+	fmt.Printf("add HTLC %d to chan %d with data %x and RHash %x\n", args.Amt, args.ChanIdx, args.Data, args.RHash)
```
Comment (gertjaap) [inline litrpc/chancmds.go]: `log.Printf` ?

File: qln/buildtx.go [hunk 5]
```diff
@@ -191,7 +210,74 @@ func (q *Qchan) BuildStateTx(mine bool) (*wire.MsgTx, error) {
 	outFancy := wire.NewTxOut(fancyAmt, fancyScript)
 	outPKH := wire.NewTxOut(pkhAmt, pkhScript)
 
-	log.Printf("\tcombined refund %x, pkh %x\n", pkhPub, outPKH.PkScript)
+	fmt.Printf("\tcombined refund %x, pkh %x, amt %d\n", pkhPub, outPKH.PkScript, pkhAmt)
```
Comment (gertjaap) [inline qln/buildtx.go]: `log.Printf` ?

File: qln/signtx.go [hunk 6]
```diff
@@ -148,70 +151,152 @@ func (nd *LitNode) SignClaimTx(claimTx *wire.MsgTx, value int64, pre []byte,
...
 	if err != nil {
-		return sig, err
+		return sig, nil, err
 	}
 
-	log.Printf("____ sig creation for channel (%d,%d):\n", q.Peer(), q.Idx())
-	log.Printf("\tinput %s\n", tx.TxIn[0].PreviousOutPoint.String())
-	for i, txout := range tx.TxOut {
-		log.Printf("\toutput %d: %x %d\n", i, txout.PkScript, txout.Value)
+	fmt.Printf("____ sig creation for channel (%d,%d):\n", q.Peer(), q.Idx())
```
Comment (gertjaap) [inline qln/signtx.go]: `log.Printf`? If so, maybe just global search. Most of these `fmt` statements were replaced by `log` when you were working on this.

**Comment by gertjaap** [comment]
The code now automatically claims a timeout (both on- and offchain). 

You can claim an HTLC manually when it went on-chain (so for closed channels) using `claim {preimage}`

**Comment by adiabat** [comment]
Noted bugs to fix:
fees are too high when adding HTLCs; the HTLC output amount shouldn't change.

also add failures.  All of em.  And have them pass up to the RPC layer so you can get a JSON error back to the user.

**Pull request merged**